### PR TITLE
Rechunked repr, result type

### PR DESCRIPTION
**Pull request by TomAugspurger**
Early prototype of what the result type can be.

** Text repr **

Just combines the reprs of the three objects

```
<Rechunked>
* Source: <zarr.core.Array (8000, 8000) float32>
* Intermediate: dask.array<from-zarr, shape=(8000, 8000), dtype=float32, chunksize=(8000, 800), chunktype=numpy.ndarray>
* Target: Delayed('store-63c77026-c532-11ea-a7a9-186590cd1c87')
```

This is simple, but can be improved:

1. The "target" repr is meaningless
2. We've lost information about storage (gcs, s3, etc.)
3. Do users care that the intermediate repr is a Dask Array (no)?  They perhaps care about the shape & chunk shape of the intermediate array. 

** HTML repr**

Collapsed

Screen Shot 2020-07-13 at 1 09 25 PM

Expanded

Screen Shot 2020-07-13 at 1 09 35 PM

Again, going for simplicity of implementation so far. The `target` repr is still meaningless. Need to think a bit about how to better display this information.

**High-level questions**

1. What official names should we use for the three stages. I've used "source", "intermediate", and "target".

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Forgot to add a discussion item about what the result of `result.compute()` should actually be. Right now it's a `List[None]`, one per chunk, which doesn't seem useful. Playing with some options now.

**Comment by TomAugspurger** [comment]
@rabernat in general, I think the result type should match the input type. So if you do `rechunker.rechunk(zarr.Array)` you get back a Zarr Array. Likewise for Dask Array, xarray.Dataset, etc.

If we agree with that principle, then I think we should consider two other things

1. Have `rechunk` be eager, rather than delayed. Having `rechunk()` always return a Delayed result can make things strange: `rechunker.rechunk(...).compute()` can returns a Dask Array, which is a lazy object. That feels a bit wrong to me.
2. Expose some kind of `rechunker.plan_rechunk()` that does things lazily. This is where we would put the nice repr. The result would be something like `Delayed`, and would have a `.execute()` or `.compute()` method that would kick off the computation and get the result.

**Comment by rabernat** [comment]
Thanks for thinking this over. I slightly favor option 2. I think it is good to force the user to examine their plan before computing.

**Comment by rabernat** [comment]
> 1. What official names should we use for the three stages. I've used "source", "intermediate", and "target".

:+1:

**Comment by TomAugspurger** [comment]
FWIW, I'm going back and forth. I'm mostly uncomfortable with `<thing>.compute()` returning a lazy object like a Dask Array. Maybe we just keep the current API and throw on a `Rechunked.execute()` method that's an alias for compute.

**Comment by TomAugspurger** [comment]
> Maybe we just keep the current API and throw on a Rechunked.execute() method that's an alias for compute.

@rabernat this feels really nice. The usage now is roughly

```python
plan = rechunker.rechunk(zarr.array.Array, ...)
plan  # inspect the plan. We can include any useful information here.
result = plan.execute()
```

and you get back the result, in this case a zarr.array.Array, matching the input type of `source_array`. We'll need to do a bit of work to make this actually be the case when we add support for other types, but I think this is good.

Here are the updated reprs.

text:

```
<Rechunked>
* Source      : <zarr.core.Array (8000, 8000) float32>
* Intermediate: dask.array<from-zarr, shape=(8000, 8000), dtype=float32, chunksize=(8000, 800), chunktype=numpy.ndarray>
* Target      : <zarr.core.Array (8000, 8000) float32>
```

HTML:

Screen Shot 2020-07-13 at 2 39 56 PM

**Comment by rabernat** [comment]
This sounds great.

One hitch is how to deal with groups. Over in #13 I am working on making rechunk accept zarr groups. This complicates things a bit.

**Comment by TomAugspurger** [comment]
Almost have something here that works with the new input types (zarr Group). Will hopefully have something within the hour.

**Comment by TomAugspurger** [comment]
Should be working, if you want to try it out.

I'd like the repr to somewhere include a concise summary of the rechunk plan

```
(source_chunks,) -> (intermediate_chunks,) -> (target_chunks,)
```

I'm not sure yet how to do that for groups.

File: tests/integration_test.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    dtype = "f4"
+
+    a_source = zarr.ones(shape, chunks=source_chunks, dtype=dtype, store=store_source)
+
+    target_store = str(tmp_path / "target.zarr")
+    temp_store = str(tmp_path / "temp.zarr")
+    max_mem = 25600000
+    target_chunks = (8000, 200)
+    return (
+        api.rechunk_zarr2zarr_w_dask(
```
Comment (rabernat) [inline tests/integration_test.py:23]: The name of this function has changed to `_rechunk_array`. Or maybe you want to use the public function `rechunk`?

Comment (rabernat) [reply tests/integration_test.py:23]: Actually, maybe you want to move all your tests to the `test_rechunk.py` module?

Comment (TomAugspurger) [reply tests/integration_test.py]: Yeah I messed up the git merge. `test_rechunk` should have everything now.

File: rechunker/api.py [hunk 2]
```diff
@@ -10,6 +13,84 @@
 from rechunker.algorithm import rechunking_plan
 
 
+class Rechunked(Delayed):
```
Comment (TomAugspurger) [inline rechunker/api.py:16]: Long-term, we might not want to subclass Dask Delayed, and instead make our own little light collection:

We don't necessarily want all the magic of `dask.delayed` creating new task graphs on `.<arbitrary_attribute>`. And if you do anything else to the Rechunked result, I think you get back a `Delayed` object, which won't have `execute`.

File: rechunker/api.py [hunk 3]
```diff
@@ -81,7 +174,28 @@ def rechunk(
             )
             stores_delayed.append(delayed)
 
-        return stores_delayed
+        # This next block makes a task that
```
Comment (TomAugspurger) [inline rechunker/api.py:177]: Hopefully this makes sense. LMK if I can clarify anything.

File: rechunker/api.py [hunk 4]
```diff
@@ -153,6 +265,13 @@ def _rechunk_array(
         target_store_delayed = dsa.store(
             source_read, target_array, lock=False, compute=False
         )
+        return Rechunked(
+            target_store_delayed.key,
+            target_store_delayed.dask,
+            source=source_array,
+            intermediate=None,
```
Comment (TomAugspurger) [inline rechunker/api.py:272]: Will need to think how to represent this in the repr.

File: rechunker/api.py [hunk 5]
```diff
@@ -10,6 +13,84 @@
...
+        """
+        self.compute(**kwargs)
+        return self._target
+
+    def __repr__(self):
+        return textwrap.dedent(
+            f"""\
+            <Rechunked>
+            * Source      : {repr(self._source)}
+            * Intermediate: {repr(self._intermediate)}
```
Comment (rabernat) [inline rechunker/api.py:54]: I would put this in an if block and not even show intermediate if it is None.

File: rechunker/api.py (hunk 3 above)
Comment (rabernat) [reply rechunker/api.py:177]: It's a bit above my pay grade! Just trying to learn what is going on here...

File: rechunker/api.py (hunk 2 above)
Comment (rabernat) [reply rechunker/api.py:16]: Would we want this class documented in the docs?

File: tests/test_rechunk.py [hunk 6]
```diff
@@ -88,6 +90,7 @@ def test_rechunk_group(tmp_path):
     delayed = api.rechunk(
         group, target_chunks, max_mem, target_store, temp_store=temp_store
     )
+    assert isinstance(delayed, api.Rechunked)
```
Comment (rabernat) [inline tests/test_rechunk.py:93]: A few lines below we call `dask.compute(delayed)`, which appears to do the right thing. But should we call `execute()` instead?

Comment (TomAugspurger) [reply tests/test_rechunk.py:93]: Will take a look (probably tomorrow). I'd recommend execute, but they'll be the same if the value in the dask graph for `delayed.dask[delayed.key]` is the target array.

Comment (TomAugspurger) [reply tests/test_rechunk.py:93]: I'm still thinking through this. If we return a zarr.Array here I'll need to revisit https://github.com/dask/dask/issues/5105#issuecomment-625282112 and https://github.com/dask/dask/pull/6203. Just won't have time today, or probably this week.

Anyway I would recommend using `.execute()` for now. They're equivalent as far as doing the rechunking goes, it's just that `.execute()` returns the `target` array, which may be more useful.

**Comment by TomAugspurger** [comment]
This should be good to go.

**Pull request merged**